File: SolutionTest/langx_test.cpp
#include "langX.h"

#include <iostream>
#include <filesystem>
#include <chrono>
#include <regex>
#include <fstream>
#include <iomanip>
#include <sstream>

namespace fs = std::filesystem;
using Clock = std::chrono::steady_clock;
using Sec = std::chrono::duration<double>;

static const fs::path SCRIPT_DIR = "d:/School/RTU/LangX/SolutionTest";
static const fs::path MODEL_DIR = SCRIPT_DIR / "model";
static const fs::path MODEL_PATH = MODEL_DIR / "Qwen_Qwen2.5-VL-7B-Instruct-Q4_K_M.gguf";
static const fs::path MMPROJ = MODEL_DIR / "mmproj-Qwen_Qwen2.5-VL-7B-Instruct-f16.gguf";
static const fs::path IMAGE_PATH = SCRIPT_DIR / "robot2.jpg";

static const std::string INTRO_FACT = "My name is Jimmy";
static const std::string INTRO_PROMPT =
    "Hello! I want to have a conversation with you. "
    "Here is an important fact to remember: " + INTRO_FACT + ". "
    "Now, what color is the sky on a clear day?";

struct TestPrompt {
    std::string name;
    std::string prompt;
    std::string image;
    std::string pattern;
    std::string desc;
};

static bool check(const std::string& response, const std::string& pattern) {
    return std::regex_search(response, std::regex(pattern, std::regex_constants::icase));
}

static std::string timestamp_str() {
    auto t = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    std::tm tm;
    localtime_s(&tm, &t);
    std::ostringstream o;
    o << std::put_time(&tm, "%Y%m%d_%H%M%S");
    return o.str();
}

static std::string iso_now() {
    auto t = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    std::tm tm;
    localtime_s(&tm, &t);
    std::ostringstream o;
    o << std::put_time(&tm, "%Y-%m-%dT%H:%M:%S");
    return o.str();
}

int main() {
    const std::vector<TestPrompt> prompts = {
        {"introduction", INTRO_PROMPT, "", "blue", "Basic question about sky color"},
        {"image_describe", "Describe what you see in this image in detail.", IMAGE_PATH.string(), "robot|mech|machine|armor|metal|figure|character", "Describe the robot2.jpg image"},
        {"fact_recall", "Earlier I told you an important fact. What was it? Repeat it.", "", "Jimmy", "Recall fact from the introduction"},
    };

    fs::path out_file = SCRIPT_DIR / ("result_langx_cpp_" + timestamp_str() + ".txt");
    std::ostringstream log;

    struct Result { std::string name, status, reason; double time = 0; };
    std::vector<Result> results;

    log << "LangX C++ Test \xe2\x80\x93 " << iso_now() << "\n";
    log << "Model: " << MODEL_PATH.string() << "\n";
    log << "Image: " << IMAGE_PATH.string() << "\n";
    log << std::string(60, '=') << "\n";

    auto t_total = Clock::now();

    // --- Setup: config, model, stack ---
    langX::Config cfg;
    langX::initialize_langX(cfg);

    langX::ModelParams mp(MODEL_PATH.string(), MMPROJ.string());
    mp.n_ctx = 8192;

    std::cout << "Loading VLM model...\n";
    auto t_load = Clock::now();
    langX::initModel(mp, "vlm");
    double load_s = Sec(Clock::now() - t_load).count();
    std::cout << "Model loaded.\n\n";
    log << "Model load time: " << std::fixed << std::setprecision(2) << load_s << "s\n";

    langX::setModelSystemPrompt("You are a helpful assistant. Answer concisely.");

    langX::InquerySettings settings;
    settings.seed = langX::randomSeed();

    langX::Stack* stack = langX::makeDefaultStack("vlm_test");
    langX::setInquerySettings(settings, stack);
    langX::switchModel("vlm", stack);
    langX::initConversation("test_convo", stack);

    // --- Test prompt loop ---
    for (const auto& p : prompts) {
        std::cout << "\n" << std::string(60, '=') << "\n";
        std::cout << "[" << p.name << "] " << p.desc << "\n";
        std::cout << "Prompt: " << p.prompt << "\n";
        std::cout << std::string(60, '-') << "\n";

        log << "\n--- " << p.name << ": " << p.desc << " ---\n";
        log << "Prompt: " << p.prompt << "\n";

        std::vector<std::string> files;
        if (!p.image.empty()) files.push_back(p.image);

        auto t_prompt = Clock::now();
        std::string response = langX::inference(stack, p.prompt.c_str(), files);
        double prompt_s = Sec(Clock::now() - t_prompt).count();

        std::cout << response << "\n";

        bool passed = check(response, p.pattern);
        std::string status = passed ? "PASS" : "FAIL";
        results.push_back({p.name, status, "", prompt_s});

        log << "Response: " << response << "\n";
        log << "Pattern: " << p.pattern << "\n";
        log << "Result: " << status << " (" << std::fixed << std::setprecision(2) << prompt_s << "s)\n";

        std::cout << "[" << status << "] pattern=/" << p.pattern << "/\n";
    }

    // --- Summary ---
    log << "\n" << std::string(60, '=') << "\n";
    log << "SUMMARY\n";
    int total = (int)results.size();
    int n_pass = 0, n_fail = 0, n_err = 0;
    for (auto& r : results) {
        if (r.status == "PASS") n_pass++;
        else if (r.status == "FAIL") n_fail++;
        else n_err++;
    }

    for (auto& r : results) {
        log << "  " << r.name << ": " << r.status;
        if (r.time > 0) log << " (" << std::fixed << std::setprecision(2) << r.time << "s)";
        log << "\n";
    }

    double total_s = Sec(Clock::now() - t_total).count();
    log << "\nTotal: " << n_pass << "/" << total << " passed, "
        << n_fail << " failed, " << n_err << " errors\n";
    log << "Total time: " << std::fixed << std::setprecision(2) << total_s << "s\n";

    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "SUMMARY: " << n_pass << "/" << total << " passed, " << n_fail << " failed, " << n_err << " errors\n";
    std::cout << "Results saved to: " << out_file.string() << "\n";

    std::ofstream(out_file) << log.str();
    return 0;
}

File: SolutionTest/langx_test.py
import sys, os, re, traceback, time
from datetime import datetime
# local python library
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
LANGX_DIR  = os.path.join(SCRIPT_DIR, "..", "llama_langx", "langXpy", "langx")
sys.path.insert(0, LANGX_DIR)
import langx_core as lx

MODEL_DIR = os.path.join(SCRIPT_DIR, "model")
MODEL_PATH = os.path.join(MODEL_DIR, "Qwen_Qwen2.5-VL-7B-Instruct-Q4_K_M.gguf")
MMPROJ = os.path.join(MODEL_DIR, "mmproj-Qwen_Qwen2.5-VL-7B-Instruct-f16.gguf")
IMAGE_PATH = os.path.join(SCRIPT_DIR, "robot2.jpg")

INTRO_FACT = "My name is Jimmy"
INTRO_PROMPT = (
    f"Hello! I want to have a conversation with you. "
    f"Here is an important fact to remember: {INTRO_FACT}. "
    f"Now, what color is the sky on a clear day?"
)

PROMPTS = [
    {
        "name": "introduction",
        "prompt": INTRO_PROMPT,
        "image": None,
        "pattern": r"blue",
        "desc": "Basic question about sky color",
    },
    {
        "name": "image_describe",
        "prompt": "Describe what you see in this image in detail.",
        "image": IMAGE_PATH,
        "pattern": r"robot|mech|machine|armor|metal|figure|character",
        "desc": "Describe the robot2.jpg image",
    },
    {
        "name": "fact_recall",
        "prompt": "Earlier I told you an important fact. What was it? Repeat it.",
        "image": None,
        "pattern": r"Jimmy",
        "desc": "Recall fact from the introduction",
    },
]

def check(response: str, pattern: str) -> bool:
    return bool(re.search(pattern, response, re.IGNORECASE))


def main():
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    out_file = os.path.join(SCRIPT_DIR, f"result_langx_py_{timestamp}.txt")
    results = []

    log_lines = [f"LangX Python Test – {datetime.now().isoformat()}\n"]
    log_lines.append(f"Model: {MODEL_PATH}\n")
    log_lines.append(f"Image: {IMAGE_PATH}\n")
    log_lines.append("=" * 60 + "\n")

    t_total = time.perf_counter()
    try:
        # --- Preparing the model: config, load and system prompt ---
        cfg = lx.Config()
        lx.initialize_langX(cfg)
        mp = lx.ModelParams(MODEL_PATH, MMPROJ)
        mp.n_ctx = 8192 # needed for image
        print("Loading VLM model...")
        t_load = time.perf_counter()
        lx.init_model(mp, "vlm")
        t_load = time.perf_counter() - t_load
        print("Model loaded.\n")
        log_lines.append(f"Model load time: {t_load:.2f}s\n")

        lx.set_model_system_prompt("You are a helpful assistant. Answer concisely.")

        # --- Preparing the conversation: settings, random seed ---
        settings = lx.InquerySettings()
        settings.seed = lx.random_seed()

        # --- Preparing the langX stack ---
        stack = lx.make_default_stack("vlm_test")
        lx.set_inquery_settings(settings, stack)
        lx.switch_model("vlm", stack)
        lx.init_conversation("test_convo", stack)
    except Exception as e:
        msg = f"SETUP FAILED: {e}\n{traceback.format_exc()}"
        print(msg)
        log_lines.append(msg)
        for p in PROMPTS:
            results.append({"name": p["name"], "status": "FAILED", "reason": "setup error"})
            log_lines.append(f"\n--- {p['name']}: FAILED (setup error) ---\n")
        with open(out_file, "w", encoding="utf-8") as f:
            f.writelines(log_lines)
        return

    # --- Test Prompt Loop ---
    for p in PROMPTS:
        print(f"\n{'='*60}")
        print(f"[{p['name']}] {p['desc']}")
        print(f"Prompt: {p['prompt']}")
        print("-" * 60)

        log_lines.append(f"\n--- {p['name']}: {p['desc']} ---\n")
        log_lines.append(f"Prompt: {p['prompt']}\n")

        try:
            files = [p["image"]] if p["image"] else []
            t_prompt = time.perf_counter()
            response = lx.inference(stack, p["prompt"], files)
            t_prompt = time.perf_counter() - t_prompt
            print(response)

            passed = check(response, p["pattern"])
            status = "PASS" if passed else "FAIL"
            results.append({"name": p["name"], "status": status, "time": t_prompt})

            log_lines.append(f"Response: {response}\n")
            log_lines.append(f"Pattern: {p['pattern']}\n")
            log_lines.append(f"Result: {status} ({t_prompt:.2f}s)\n")

            print(f"[{status}] pattern=/{p['pattern']}/")
        except Exception as e:
            tb = traceback.format_exc()
            results.append({"name": p["name"], "status": "FAILED", "reason": str(e)})
            log_lines.append(f"ERROR: {e}\n{tb}\n")
            log_lines.append("Result: FAILED (exception)\n")
            print(f"[FAILED] {e}")

    # --- Summary ---
    log_lines.append("\n" + "=" * 60 + "\n")
    log_lines.append("SUMMARY\n")
    total = len(results)
    passed = sum(1 for r in results if r["status"] == "PASS")
    failed = sum(1 for r in results if r["status"] == "FAIL")
    errors = sum(1 for r in results if r["status"] == "FAILED")

    for r in results:
        t = f" ({r['time']:.2f}s)" if "time" in r else ""
        log_lines.append(f"  {r['name']}: {r['status']}{t}\n")
    t_total = time.perf_counter() - t_total
    log_lines.append(f"\nTotal: {passed}/{total} passed, {failed} failed, {errors} errors\n")
    log_lines.append(f"Total time: {t_total:.2f}s\n")

    print(f"\n{'='*60}")
    print(f"SUMMARY: {passed}/{total} passed, {failed} failed, {errors} errors")
    print(f"Results saved to: {out_file}")

    with open(out_file, "w", encoding="utf-8") as f:
        f.writelines(log_lines)

if __name__ == "__main__":
    main()
